New package with automatic collection transformers for Backbone

jgonggrijp opened the pull request:
This is the next NPM package in the series of outfactorings from the READ-IT interface. Previous iterations were [backbone-util](https://github.com/CentreForDigitalHumanities/backbone-util/pull/1) and before that, [timing-util](https://github.com/CentreForDigitalHumanities/js-timing-util/pull/1).

Rather than writing a story here about the purpose of the package, I'm just going to refer to the README, specifically to the [rendered version](https://github.com/CentreForDigitalHumanities/backbone-collection-transformers/blob/3fc171057445e8b0d6b62fd4bff5d9982238a04d/README.md).

@ar-jan If I'm right, you don't regularly use Backbone. That's fine, I think you will still be able to meaningfully review this package. All you need to know is that a `Backbone.Model` is essentially a wrapper around a JSON object and a `Backbone.Collection` is basically a wrapper around an array of models. Both wrappers have convenience methods, know how to synchronize with a backend and emit events so you can observe changes.

If you still want a slightly more elaborate introduction before you start, I suggest reading the sections [Getting Started](https://backbonejs.org/#Getting-started), [Models and Views](https://backbonejs.org/#Model-View-separation) and [Collections](https://backbonejs.org/#Model-Collections) from the official documentation. You will encounter more links to the Backbone documentation during your review.

It's quite a lot for a single review. Sorry about that. I tried distributing the work over multiple people with backbone-util, but it resulted in a lot of double work. On the bright side, you can take your time because I will be on vacation for a few weeks.

I suggest the following review procedure:

1. Read the rendered version of the README.
2. For each X in `collection-proxy`, `inheritance`, `filtered-collection` and `mapped-collection` (roughly in that order):
    1. Read the "rich diff" or the [rendered version](https://github.com/CentreForDigitalHumanities/backbone-collection-transformers/tree/3fc171057445e8b0d6b62fd4bff5d9982238a04d/doc) of `doc/X.md`.
    2. Read the `describe` and `it` strings in `src/X.test.js` and check whether they make intuitive sense to you.
    3. Check `src/X.js` for anything that needs clarification or that looks wrong to you.
    4. Check that the contents of `src/X.d.ts` align with those of `src/X.js`. There is no `src/inheritance.d.ts`, this is intentional.
3. Scan `src/.npmignore`, `.gitignore`, `.editorconfig`, `index.js`, `index.test.js`, `karma.conf.cjs` and the `package.json` for any choices that you find disagreeable.
4. Try running the tests. Issue `npm run test -- --single-run` and then open http://localhost:9876/ in a browser. The output will appear in your terminal.

The CONTRIBUTING and the LICENSE were copied without modification from backbone-util, where they already underwent review, so you can skip them if you want.

CC FYI @tijmenbaarda @JeltevanBoheemen @paulfalgout @stephenkao @GammaGames @ogonkov @Rayraz

jgonggrijp committed:
Bootstrap empty library based on structure of backbone-util

jgonggrijp committed:
Outfactor CollectionProxy from readit-interface

jgonggrijp committed:
Outfactor FilteredCollection from readit-interface

jgonggrijp committed:
Outfactor MappedCollection from readit-interface

jgonggrijp committed:
Add backbone-util as a dependency (git-based for now)

jgonggrijp committed:
Commit the package-lock.json

jgonggrijp committed:
Reorder karma-rollup plugins

jgonggrijp committed:
Implement a class derivation compatibility bridge

jgonggrijp committed:
Test that deriveConstructor creates a valid prototype chain by itself

jgonggrijp committed:
Fix inheritance prototype chain

jgonggrijp committed:
Implement parent function to emulate super

jgonggrijp committed:
Document CollectionProxy

jgonggrijp committed:
Document FilteredCollection

jgonggrijp committed:
Document MappedCollection

jgonggrijp committed:
Document inheritance

jgonggrijp committed:
Populate the index.js

jgonggrijp committed:
Complete the README

jgonggrijp committed:
Add transform tag to package.json

jgonggrijp committed:
Upgrade to the released backbone-util

jgonggrijp committed:
Bump the version to 0.1.0

ar-jan commented on doc/collection-proxy.md:
````diff
@@ -0,0 +1,27 @@
+
+```typescript
+import { mixin } from '@uu-cdh/backbone-util';
+import { ProxyMixin } from '@uu-cdh/backbone-collection-transformers';
````
Should this be `import { CollectionProxy as ProxyMixin }`? index.js has `export { default as CollectionProxy } from './src/collection-proxy.js';`.

ar-jan commented on README.md:
```diff
@@ -0,0 +1,134 @@
+// ]
+
+// Both layers automatically change along with changes in the input.
+chinesActors.set({id: 2, familyName: 'Li'});
```
typo: chineseActors

ar-jan commented on README.md:
```diff
@@ -0,0 +1,134 @@
+// Both layers automatically change along with changes in the input.
+chinesActors.set({id: 2, familyName: 'Li'});
+// joinedNameActors now also contains {id: 2, fullName: 'Li Lianjie'}
+chinesActors.remove(3);
```
chineseActors

ar-jan commented on src/mapped-collection.js:
```diff
@@ -0,0 +1,347 @@
+function ctorStart(underlying, conversion, options) {
+    options = extend({convert: wrapModelIteratee(conversion)}, options);
+    return [underlying.models, options];
+}
```
The last line of the example code in the READMe `joinedNameActors.underlying; // chineseActors` results in `TypeError: Cannot read properties of undefined (reading 'underlying') at get underlying [as underlying]`.
As we discovered, convert is missing the `underlying` argument.

ar-jan commented on src/filtered-collection.d.ts:
```diff
@@ -0,0 +1,35 @@
+export default function deriveFiltered<
+    M extends Model = Model,
+    U extends Collection<M> = Collection<M>,
+    B extends collection<M> = Collection<M>
```
Capitalization: `B extends Collection<M> = Collection<M>`

ar-jan commented on src/collection-proxy.d.ts:
```diff
@@ -0,0 +1,14 @@
+    readonly underlying: any;
+}
+
+export default const CollectionProxy: CollectionProxy;
```
My editor says `Expression expected`. Should this be:
```
declare const CollectionProxy: CollectionProxy;
export default CollectionProxy;
```
?

May be just a strict mode quirk.

ar-jan commented on src/filtered-collection.d.ts:
```diff
@@ -0,0 +1,35 @@
+
+interface FilteredCollection<
+    M extends Model = Model,
+    U extends Collection<M> = Collection<M>
```
Missing comma

ar-jan commented on src/filtered-collection.d.ts:
```diff
@@ -0,0 +1,35 @@
+> extends CollectionProxy<M, U> {}
+declare class FilteredCollection<
+    M extends Model = Model,
+    U extends Collection<M> = Collection<M>
```
Missing comma

ar-jan commented on src/mapped-collection.d.ts:
```diff
@@ -0,0 +1,57 @@
+    // Redeclaring the type of the add method in order to accomodate AddOptions.
+    add(model: {} | T, options?: AddOptions): T & Traceable;
+    add(models: Array<{} | T>, options?: AddOptions): (T & Traceable)[];
+}
```
Missing newline

jgonggrijp replied:
This is intentional. The same type is first declared as an interface and then as a class, in order to merge the declarations. However, if you still think there should be a blank line in between, I will add it here and also in the other places where I did this.

ar-jan replied:
In that case just leave it as-is!

ar-jan commented on src/collection-proxy.js:
```diff
@@ -0,0 +1,26 @@
+     */
+
+    /**
+     * Clients can this getter to transparently retrieve the innermost
```
Clients can **use** ...

ar-jan reviewed (changes requested):
The library looks carefully designed and organized, is very well documented, and contains extensive tests. The thorough background and documentation on inheritance is fascinating in its own, and was helpful for someone like me who isn't very familiar with the internals of JavaScript.

Besides some minor things in the review comments, the main thing to address is to fix the `.underlying` getter for MappedCollection, and add tests for this functionality.
I was also a bit confused by the naming in the usage of `ProxyMixin` vs `CollectionProxy`, and about `ProxyCollection` in the docs vs `CollectionProxy` in code. Synchronize the names used, or call out the choices for them more explicitly?

jgonggrijp committed:
Fix typo: chines(e)Actors

Credit to @ar-jan for noticing this in #1.

jgonggrijp committed:
Add a missing verb

jgonggrijp committed:
Rename CollectionProxy to ProxyMixin and use this name consistently

Credit to @ar-jan in #1 for pointing out the former inconsistency.

jgonggrijp committed:
Ignore emacs backup files

jgonggrijp committed:
ProxyMixin is just an object, no need to take its prototype

jgonggrijp committed:
Clarify that ProxyMixin also works for non-collections

jgonggrijp committed:
Test for proper .underlying reference, fix MappedCollection

Thanks to @ar-jan for pointing out this bug in #1.

jgonggrijp committed:
Fix typo in filtered-collection.d.ts (thanks @ar-jan in #1)

jgonggrijp committed:
Work around TypeScript strictness (thanks @ar-jan in #1)

jgonggrijp committed:
Add missing commas (thanks @ar-jan in #1)

jgonggrijp commented:
Thank you for your thorough review, @ar-jan! I think your comments helped to improve the quality of the package considerably.

If you don't mind, please give this a quick review to check that I addressed your comments sufficiently. I renamed `CollectionProxy` to `ProxyMixin` (which I already used in a few places) because I think the latter names is more accurate. There is still `ProxyCollection`, but this is the name of a collection that uses the mixin.

ar-jan commented:
All looks good! Except you reintroduced `CollectionProxy` in proxy-mixin.d.ts.

jgonggrijp committed:
Restore ProxyMixin in .d.ts (fixup 30af70c)

jgonggrijp commented:
Good catch, thanks!

jgonggrijp committed:
.npmignore emacs backup files

jgonggrijp commented:
Now [released to NPM](https://www.npmjs.com/package/@uu-cdh/backbone-collection-transformers).
